### ISSUE TITLE
Resource-management for resumptions

### ISSUE THREAD

**Issue by gadmm** [dhil/ocaml-multicont#3]
As a follow-up to the discussion at https://discuss.ocaml.org/t/multi-shot-continuations-gone-forever/9072/13, I think the usual resource-management issues inside resumptions apply to resumptions themselves: currently they have to be dropped manually, which prevents us from referring to them inside other resumptions since it becomes impossible to track when they should be freed.

As suggested in the discussion, one solution is to make resumptions purely GC-managed values. To do so, one can represent a resumption either as a heap value obtained by copying a fiber, or as a fiber which is only discontinued by the GC using a finaliser and which is never reinstated/converted back to a continuation. The interface gets simplified by removal of `demote`, `drop` and `Resumption_already_dropped`.

The distinction between linear continuations, which are allowed to own resources, and multi-shot continuations, which are not, could be explained in the documentation, to underline the issues arising from promoting arbitrary continuations.

(cc @gasche)

**Comment by gasche** [comment]
I agree, and/but I suspect that sending a PR to propose the new implementation may be more effective, in case @dhil rather prefers to keep this as a low-effort project on the side.

**Comment by gadmm** [comment]
Let me know what you think @dhil!

**Comment by dhil** [comment]
I think it is a good idea! I am interested in exploring such designs. My initial plan was just to replicate `clone_continuation`, such that I can continue working on my research software which is written in an older version of Multicore OCaml.

**Comment by dhil** [comment]
@gadmm I have managed to get byte code compilation going now. I will look into experimenting with GC'ed resumptions next.

**Comment by dhil** [comment]
Commits 2ac61580024bf583465171710e85f2f868f6b8ba and 1f7fc2653d996c6ed80b7e4a30dd853fb38fe3c9 add initial support for experimenting with GC managed resumptions/continuations.
